### PR TITLE
bundleRecommendations component

### PR DESCRIPTION
**Pull request by chrisFrazier77**
new component for upcoming bundlerecs feature

### CODE REVIEW

File: packages/snap-preact-components/src/components/Organisms/BundledRecommendations/BundledRecommendations.tsx [hunk 1]
```diff
@@ -0,0 +1,515 @@
...
+	loop?: boolean;
+	pagination?: boolean;
+	resultComponent?: JSX.Element;
+	ctaSlot?: JSX.Element;
+	preselectedCount?: number;
+	showQuantityPicker?: boolean;
+	showCheckboxes?: boolean;
+	seedInCarousel?: boolean;
+	seedText?: string;
+	seedIconOnly?: boolean;
```
Comment (korgon) [inline packages/snap-preact-components/src/components/Organisms/BundledRecommendations/BundledRecommendations.tsx:501]: I wonder if something more like `hideCarouselSeparatorIcons` or similar would be more clear? Or maybe just rename to `seedSeparatorIconOnly`?

File: packages/snap-preact-components/src/components/Organisms/BundledRecommendations/BundledRecommendations.tsx
Comment (korgon) [inline packages/snap-preact-components/src/components/Organisms/BundledRecommendations/BundledRecommendations.tsx:1]: The original component is singular `Recommendation` - perhaps this one should be as well (`BundledRecommendation`)?

File: packages/snap-preact-components/src/types.ts [hunk 2]
```diff
@@ -38,6 +38,9 @@ export type BreakpointsProps = {
 
 export type BreakpointsEntry = {
 	[property: string]: any;
+	slidesPerView?: number;
+	slidesPerGroup?: number;
+	spaceBetween?: number;
```
Comment (korgon) [inline packages/snap-preact-components/src/types.ts:43]: This makes this type specific to carousel, previous typing should have supported any number of properties. Was this necessary?

File: packages/snap-preact-demo/src/components/Recommendations/Recs/Recs.tsx
Comment (korgon) [inline packages/snap-preact-demo/src/components/Recommendations/Recs/Recs.tsx:1]: I experimented with this and noticed the following:

1. when using `loop` selecting products sometimes results in the carousel shifting position (go back one page then select a product to replicate)
2. when using `preselectedCount: 0` the CTA has a `0` where the strikePrice would be
3. when using quantity selector I can set it to `0` but it isn't excluded from CTA total
4. `pagination` seems broken - doesn't work quite right

File: packages/snap-preact-components/src/components/Organisms/BundledRecommendations/BundledRecommendations.tsx [hunk 3]
```diff
@@ -0,0 +1,515 @@
...
+		resultsToRender?.forEach((result, idx) => {
+			if (idx < _preSelectedCount) {
+				initialSelectedIds.push(result.id);
+			}
+		});
+	}
+
+	const [selectedIds, setSelectedIds] = useState<string[]>(initialSelectedIds);
+
+	const selectedItems = selectedIds.map((id) => ({
```
Comment (korgon) [inline packages/snap-preact-components/src/components/Organisms/BundledRecommendations/BundledRecommendations.tsx:268]: After thinking about this more, I think it makes more sense to have this be an array of products (a reference to the full product from the store). This works better for a few reasons:

- future proof (all properties are available)
- the quantity is already in the product store
- consistency - not adding some additional object (in this case one with simply `id` and `quantity`

**Review by korgon — CHANGES_REQUESTED** [review]
Good progress on this!

File: packages/snap-controller/src/Recommendation/RecommendationController.ts [hunk 4]
```diff
@@ -206,6 +212,34 @@ export class RecommendationController extends AbstractController {
 					return event;
 				},
 			},
+			addBundleToCart: (e: MouseEvent, results: selectedItem[], price: number): BeaconEvent | undefined => {
```
[the remark this answers is not in the document]
Comment (chrisFrazier77) [reply packages/snap-controller/src/Recommendation/RecommendationController.ts]: 👍

File: packages/snap-preact-components/src/components/Molecules/Carousel/Carousel.tsx [hunk 5]
```diff
@@ -298,7 +318,7 @@ export const Carousel = observer((properties: CarouselProps): JSX.Element => {
 
 				<Swiper
 					centerInsufficientSlides={true}
-					onInit={(swiper) => {
+					onBeforeInit={(swiper) => {
```
[the remark this answers is not in the document]
Comment (chrisFrazier77) [reply packages/snap-preact-components/src/components/Molecules/Carousel/Carousel.tsx:321]: onBeforeInit is needed now to set up the pagination arrows. So I kept that and moved the onInit callback back into onInit.

File: packages/snap-preact-components/src/components/Organisms/BundledRecommendations/BundledRecommendations.tsx [hunk 6]
```diff
@@ -0,0 +1,515 @@
...
+			<div {...styling} className={classnames('ss__bundled-recommendations', className)}>
+				<RecommendationProfileTracker controller={controller}>
+					{title && <h3 className="ss__bundled-recommendations__title">{title}</h3>}
+
+					<div>
+						<div
+							className={`ss__bundled-recommendations__product-wrapper ${
+								seedInCarousel
+									? 'ss__bundled-recommendations__product-wrapper-seed-in-carousel'
+									: 'ss__bundled-recommendations__product-wrapper-seed-not-in-carousel'
```
[the remark this answers is not in the document]
Comment (chrisFrazier77) [reply packages/snap-preact-components/src/components/Organisms/BundledRecommendations/BundledRecommendations.tsx]: They are both used in the styling. So I think some modifier class is needed, however I'm totally open to renaming them to something shorter.

File: packages/snap-preact-demo/src/components/Recommendations/Recs/Recs.tsx
Comment (chrisFrazier77) [reply packages/snap-preact-demo/src/components/Recommendations/Recs/Recs.tsx:1]: good catches. all should be fixed.

File: packages/snap-preact-components/src/components/Organisms/BundledRecommendations/BundledRecommendations.tsx (hunk 3 above)
Comment (chrisFrazier77) [reply packages/snap-preact-components/src/components/Organisms/BundledRecommendations/BundledRecommendations.tsx]: I like this more.

File: packages/snap-preact-components/package.json [hunk 7]
```diff
@@ -32,7 +32,7 @@
 		"dequal": "2.0.3",
 		"mobx-react-lite": "3.4.0",
 		"react-ranger": "2.1.0",
-		"swiper": "6.8.4"
+		"swiper": "9.0.5"
```
[the remark this answers is not in the document]
Comment (korgon) [reply packages/snap-preact-components/package.json:35]: We will do this in another task (if possible).

File: packages/snap-preact-components/src/components/Organisms/BundledRecommendation/BundleCTA.tsx [hunk 8]
```diff
@@ -5,10 +5,10 @@ import { observer } from 'mobx-react-lite';
 import { cloneWithProps } from '../../../utilities';
 import { Button } from '../../Atoms/Button';
 import { Price } from '../../Atoms/Price';
-import type { selectedItem } from './BundledRecommendations';
+import type { selectedItem } from './BundledRecommendation';
```
Comment (korgon) [inline packages/snap-preact-components/src/components/Organisms/BundledRecommendation/BundleCTA.tsx:8]: Type differs from `SelectedBundleItem`?

File: packages/snap-preact-components/src/components/Organisms/BundledRecommendation/BundledRecommendation.tsx [hunk 9]
```diff
@@ -145,16 +146,16 @@ export const BundledRecommendations = observer((properties: BundledRecommendatio
 		},
 	};
 
-	let props: BundledRecommendationsProps = {
+	let props: BundledRecommendationProps = {
 		// default props
 		breakpoints: JSON.parse(JSON.stringify(defaultCarouselBreakpoints)),
-		pagination: false,
+		// pagination: false,
```
Comment (korgon) [inline packages/snap-preact-components/src/components/Organisms/BundledRecommendation/BundledRecommendation.tsx:152]: Remove comment?

File: packages/snap-preact-components/src/components/Organisms/BundledRecommendation/BundleCTA.tsx [hunk 10]
```diff
@@ -17,43 +17,44 @@ export const BundledCTA = observer((properties: BundledCTAProps): JSX.Element =>
...
 });
 
 interface BundledCTAProps {
-	isMobile: boolean;
 	ctaSlot?: JSX.Element;
 	selectedItems: selectedItem[];
 	bundlePrice: number;
 	bundleStrikePrice?: number;
-	addToCartFunc: (e: any) => void;
+	onAddToCartClick: (e: any) => void;
```
Comment (korgon) [inline packages/snap-preact-components/src/components/Organisms/BundledRecommendation/BundleCTA.tsx:58]: I think we should type this properly so it is clear how it is expected to work.

File: packages/snap-preact-components/src/components/Organisms/BundledRecommendation/BundledRecommendation.tsx [hunk 11]
```diff
@@ -479,11 +466,11 @@ export const BundledRecommendations = observer((properties: BundledRecommendatio
 	);
 });
 
-export interface BundledRecommendationsProps extends ComponentProps {
+export interface BundledRecommendationProps extends ComponentProps {
 	results?: Product[];
 	controller: RecommendationController;
 	onAddToCart: (items: selectedItem[]) => void;
```
Comment (korgon) [inline packages/snap-preact-components/src/components/Organisms/BundledRecommendation/BundledRecommendation.tsx:472]: Typing should reflect new typing (`Product[]`) ?

**Review by korgon — CHANGES_REQUESTED** [review]

File: packages/snap-controller/src/Recommendation/RecommendationController.ts [hunk 12]
```diff
@@ -10,13 +10,19 @@ import type { RecommendationStore } from '@searchspring/snap-store-mobx';
 import type { Next } from '@searchspring/snap-event-manager';
 import type { RecommendationControllerConfig, BeforeSearchObj, AfterStoreObj, ControllerServices, ContextVariables } from '../types';
 
+export interface SelectedBundleItem {
```
Comment (korgon) [inline packages/snap-controller/src/Recommendation/RecommendationController.ts:13]: Let's remove this type as well.

**Review by korgon — CHANGES_REQUESTED** [review]

**Review by korgon — DISMISSED** [review]

**Comment by korgon** [comment]
Converted to draft to prevent accidental merging.

**Review by korgon — APPROVED** [review]

**Pull request merged**